File: web/src/lib/astrologer-search-filter.ts
import type { Astrologer } from "./astrologers";

/**
 * Filters the astrologer list by a free-text query (matched against name and
 * specialty) and an optional exact specialty filter. Ported from the Android
 * app's AstrologerSearchFilter.kt.
 */
export function filterAstrologers<T extends Astrologer>(
  astrologers: T[],
  query: string,
  specialty: string | null
): T[] {
  const trimmedQuery = query.trim().toLowerCase();
  return astrologers.filter((astrologer) => {
    const matchesSpecialty = specialty === null || astrologer.specialty === specialty;
    const matchesQuery =
      trimmedQuery.length === 0 ||
      astrologer.name.toLowerCase().includes(trimmedQuery) ||
      astrologer.specialty.toLowerCase().includes(trimmedQuery);
    return matchesSpecialty && matchesQuery;
  });
}

File: app/src/main/java/com/example/data/AstrologerSearchFilter.kt
package com.example.data

/**
 * Filters the astrologer list by a free-text query (matched against name and
 * specialty) and an optional exact specialty filter, extracted out of the
 * Composable so the matching rules are unit testable.
 */
object AstrologerSearchFilter {
    fun apply(astrologers: List<Astrologer>, query: String, specialty: String?): List<Astrologer> {
        return astrologers.filter { astrologer ->
            val matchesSpecialty = specialty == null || astrologer.specialty == specialty
            val matchesQuery = query.isBlank() ||
                astrologer.name.contains(query, ignoreCase = true) ||
                astrologer.specialty.contains(query, ignoreCase = true)
            matchesSpecialty && matchesQuery
        }
    }
}
